faiss_vector_posting.go

```diff
@@ -395,168 +393,112 @@ func (sb *SegmentBase) InterpretVectorIndex(field string, requiresFiltering bool
+				// filter hit IDs may be ineligible for the kNN since the document does
+				// not have any/valid vectors.
 				if len(vectorIDsToInclude) == 0 {
 					return rv, nil
 				}
-
-				// Retrieve the mapping of centroid IDs to vectors within
-				// the cluster.
-				clusterAssignment, _ := vecIndex.ObtainClusterToVecIDsFromIVFIndex()
-				// Accounting for a flat index
-				if len(clusterAssignment) == 0 {
+				// If the index is not an IVF index, then the search can be
+				// performed directly, using the Flat index.
+				if !vecIndex.IsIVFIndex() {
+					// vector IDs corresponding to the local doc numbers to be
+					// considered for the search
 					scores, ids, err := vecIndex.SearchWithIDs(qVector, k,
 						vectorIDsToInclude, params)
 					if err != nil {
 						return nil, err
 					}
-
 					addIDsToPostingsList(rv, ids, scores)
 					return rv, nil
 				}
-
-				// Converting to roaring bitmap for ease of intersect ops with
-				// the set of eligible doc IDs.
-				centroidVecIDMap := make(map[int64]*roaring.Bitmap)
-				for centroidID, vecIDs := range clusterAssignment {
-					if _, exists := centroidVecIDMap[centroidID]; !exists {
-						centroidVecIDMap[centroidID] = roaring.NewBitmap()
-					}
-					vecIDsUint32 := make([]uint32, 0, len(vecIDs))
-					for _, vecID := range vecIDs {
-						vecIDsUint32 = append(vecIDsUint32, uint32(vecID))
-					}
-					centroidVecIDMap[centroidID].AddMany(vecIDsUint32)
-				}
-
 				// Determining which clusters, identified by centroid ID,
 				// have at least one eligible vector and hence, ought to be
 				// probed.
-				eligibleCentroidIDs := make([]int64, 0)
-
+				clusterVectorCounts, err := vecIndex.ObtainClusterVectorCountsFromIVFIndex(vectorIDsToInclude)
+				if err != nil {
+					return nil, err
+				}
 				var selector faiss.Selector
-				var err error
 				// If there are more elements to be included than excluded, it
 				// might be quicker to use an exclusion selector as a filter
 				// instead of an inclusion selector.
 				if float32(len(eligibleDocIDs))/float32(len(docVecIDMap)) > 0.5 {
-					ineligibleVecIDsBitmap := roaring.NewBitmap()
-					eligibleDocIDsMap := make(map[uint64]struct{})
-					for _, eligibleDocID := range eligibleDocIDs {
-						eligibleDocIDsMap[(eligibleDocID)] = struct{}{}
+					bs := bitset.New(uint(len(eligibleDocIDs)))
+					for _, docID := range eligibleDocIDs {
+						bs.Set(uint(docID))
 					}
-
-					ineligibleVectorIDs := make([]int64, 0, len(vecDocIDMap)-
-						len(vectorIDsToInclude))
-
+					ineligibleVectorIDs := make([]int64, 0, len(vecDocIDMap)-len(vectorIDsToInclude))
 					for docID, vecIDs := range docVecIDMap {
-						if _, exists := eligibleDocIDsMap[uint64(docID)]; !exists {
-							for _, vecID := range vecIDs {
-								ineligibleVecIDsBitmap.Add(uint32(vecID))
-								ineligibleVectorIDs = append(ineligibleVectorIDs, vecID)
+						if !bs.Test(uint(docID)) {
+							if len(vecIDs) == 1 {
```

Comment:
I see the reason why you're doing this per the conversation with @Thejas-bhat above - but given how this looks I feel someone in the future may refactor this to simply appending the slice. So, would you add some commentary here and above on why we're doing the len(..) == 1 check and the else.